qutip/core/data/convert.pyx

```diff
@@ -259,63 +261,171 @@ cdef class _to:
+
+        Parameters
+        ----------
+        dtype : type, str
+            Either the (case-insensitive) name of a data-layer type or a type
+            itself.
+
+        Returns
+        -------
+        type
+            A data-layer type.
+
+        Raises
+        ------
+        TypeError
+            If ``dtype`` is neither a string nor a type.
+        ValueError
+            If ``dtype`` is a name, but no data-layer type of that name is
+            registered, or if ``dtype`` is a type, but not a known data-layer
+            type.
+        """
+        if type(dtype) is type:
+            if dtype not in self.dtypes:
+                raise ValueError(
+                    "Type is not a data-layer type: " + repr(dtype))
+            return dtype
+        elif type(dtype) is str:
+            if dtype.lower() not in self._str2type:
+                raise ValueError(
+                    "Type name is not known to the data-layer: " + repr(dtype))
+            return self._str2type[dtype.lower()]
+        raise TypeError(
+            "Invalid dtype is neither a type nor a type name: " + repr(dtype))
 
     @cython.boundscheck(False)
     @cython.wraparound(False)
     def __getitem__(self, arg):
-        if isinstance(arg, type):
+        if type(arg) is not tuple:
             arg = (arg,)
-        if not isinstance(arg, tuple) or not arg or len(arg) > 2:
+        if not arg or len(arg) > 2:
             raise KeyError(arg)
-        to_t = arg[0]
-        if to_t not in self.dtypes:
-            raise TypeError("to_type is not known: " + str(to_t))
+        to_t = self.parse(arg[0])
         if len(arg) == 1:
             converters = {
                 from_t: self._convert[to_t, from_t] for from_t in self.dtypes
             }
             return _partial_converter(converters, to_t)
-        from_t = arg[1]
-        if from_t not in self.dtypes:
-            raise TypeError("from_type is not known: " + str(from_t))
+        from_t = self.parse(arg[1])
         return self._convert[to_t, from_t]
 
     def __call__(self, to_type, data):
-        if not isinstance(to_type, type):
-            raise TypeError(repr(to_type) + " is not a type object")
-        if to_type not in self.dtypes:
-            raise ValueError("unknown output type: " + to_type.__name__)
-        from_type = type(data)
-        if from_type not in self.dtypes:
-            raise TypeError("unknown input type: " + from_type.__name__)
+        to_type = self.parse(to_type)
+        from_type = self.parse(type(data))
         if to_type == from_type:
             return data
         return self._convert[to_type, from_type](data)
 
 
 cdef class _create:
+    cdef readonly list _creators
+
     def __init__(self):
-        pass
+        self._creators = []
 
     def add_creators(self, creators):
-        pass
+        """
+        Add creation functions to make a data-layer object from an arbitrary
```

Comment:
@Ericgig I update the docstring. Let me know what you think.

@Ericgig @jakelishman Minor complaint -- this function is called `add_creators` but it completely overrides the list of creators each time, unless I am missing something? Perhaps not something we need to fix in this PR, but I imagine we will need to fix it for the two GSoC projects that are starting soon so that they can add their own creators?

Reply:
I created this function in this PR and yes allowing to add new creator for the new data-type elsewhere is worth it (and expected behaviour from the name). So I should fix it before merging, well spotted.

Reply:
Ah, just saw this comment about `add_creators` - I've got the same comment in my review as well.